test/commandbuilderharness.js

```diff
@@ -0,0 +1,161 @@
+const { expect } = require("chai");
+const { ethers } = require("hardhat");
+const weiroll = require("@weiroll/weiroll.js");
+
+async function deployLibrary(name) {
+  const factory = await ethers.getContractFactory(name);
+  const contract = await factory.deploy();
+  return weiroll.Contract.fromEthersContract(contract);
+}
+
+describe("CommandBuilderHarness", function () {
+  let cbh;
+  let math;
+  let strings;
+  let abi = ethers.utils.defaultAbiCoder;
+
+  before(async () => {
+    const Cbh = await ethers.getContractFactory("CommandBuilderHarness");
+    cbh = await Cbh.deploy();
+
+    math = await deployLibrary("Math");
+    strings = await deployLibrary("Strings");
+  });
+
+  async function executeBuildInputs(commands, state, abiout, msg){
+    for (let c of commands) {
+        selector = c.slice(0, 10);
+        indices = "0x" + c.slice(10, 10+7*2);
+        target = "0x" + c.slice(10+7*2);
+        const tx = await cbh.testBuildInputs(state, selector, indices);
```

Comment:
If you mark the `testBuildInputs` and `testWriteOutputs` functions as `view` or `pure`, you can give them return types, and call them locally and just get the return data, rather than having to create a transaction and log the output in an event.